README.md

```diff
@@ -1,707 +1,1799 @@
+  - [Why is my selector recomputing when the input state stays the same?](#why-is-my-selector-recomputing-when-the-input-state-stays-the-same)
+  - [Can I use `Reselect` without `Redux`?](#can-i-use-reselect-without-redux)
+  - [How do I create a selector that takes an argument?](#how-do-i-create-a-selector-that-takes-an-argument)
+  - [The default memoization function is no good, can I use a different one?](#the-default-memoization-function-is-no-good-can-i-use-a-different-one)
+  - [How do I test a selector?](#how-do-i-test-a-selector)
+  - [Can I share a selector across multiple component instances?](#can-i-share-a-selector-across-multiple-component-instances)
+  - [Are there TypeScript Typings?](#are-there-typescript-typings)
+  - [I am seeing a TypeScript error: `Type instantiation is excessively deep and possibly infinite`](#i-am-seeing-a-typescript-error-type-instantiation-is-excessively-deep-and-possibly-infinite)
+  - [How can I make a curried selector?](#how-can-i-make-a-curried-selector)
+  - [How can I make pre-typed version of `createSelector` for my root state?](#how-can-i-make-pre-typed-version-of-createselector-for-my-root-state)
+  - [What if I want to use `createSelector` without memoization?](#what-if-i-want-to-use-createselector-without-memoization)
+- [External References](#external-references)
+- [Related Projects](#related-projects)
+- [License](#license)
+- [Prior Art and Inspiration](#prior-art-and-inspiration)
+
+---
+
+## Terminology
+
+<details><summary><b>Click to expand</b></summary>
+
+- <a name="selector-function"></a>[**`Selector Function`**](#selector-function): A function that takes the `Redux` store state (or a part of it) as an argument and returns data derived from that state.
+- <a name="input-selectors"></a>[**`input selectors`**](#input-selectors): Basic selector functions used as building blocks for creating a memoized selector. They are passed as the first argument(s) to [`createSelector`].
+- <a name="output-selector"></a>[**`Output Selector`**](#output-selector): The actual memoized selectors created by [`createSelector`].
+- <a name="result-function"></a>[**`Result Function (Combiner)`**](#result-function): The function that comes after the [`input selectors`]. It takes the [`input selectors`]' return values as arguments and returns a result. Otherwise known as the `combiner`.
+- <a name="combiner"></a>[**`Combiner`**](#combiner): A function that takes [`input selectors`]' return values as arguments and returns a result (This function comes after the [`input selectors`] inside [`createSelector`]). This term is somewhat interchangeably used with `Result Function` or `resultFunc`. But `combiner` is more of a general term and `resultFunc` is more specific to `Reselect`. So the `resultFunc` is a `combiner` but a `combiner` is not necessarily the same as `resultFunc`. For the sake of simplicity, they are used synonymously throughout this documentation.
```

Comment:
:hand: if we're going to consolidate terms, I'd like to drop "combiner" entirely and just stick with "result function" (or possibly "output function").

Reply:
Sounds good, the only reason I included "combiner" is because of how it was referened internally. But you're right we can drop that term in the docs to avoid further confusion. I think we can just stick to "result function". "output function" can be confused with "output selector". What do you think?

Reply:
Oh. I missed the use of "output selector" here.  I don't think I like that usage tbh :)  I personally have used that phrase to mean what we're now calling the "result function".  If we have to distinguish things, I'd say "generated selector", maybe?

Reply:
Honestly I think "output selector" makes sense for the generated selector. It would also be consistent with how the types were named initially. The result function is not really a "selector" if you think about it. The way it is now we have "input selectors", then "result function" does the calculation, and we get an "output selector". What do you think?

Reply:
So it's kind of like:

input -> calculation -> output

Reply:
eh, I'm also not sold on the type names either, even though they've been there since early on :)  "Parametric selector" has always confused me.

I think I'd like to stick with either "generated selector" or "memoized selector" as the term for the function returned by `createSelector`.

Reply:
> eh, I'm also not sold on the type names either, even though they've been there since early on :) "Parametric selector" has always confused me

Tbh, me neither, but I was just trying to stay consistent with the type names. And you're right about the "parametric selector", I'm not really a fan of it either. But they were there since the beginning and I thought that's how you would want me to name things in the docs as well.

Reply:
> I think I'd like to stick with either "generated selector" or "memoized selector" as the term for the function returned by `createSelector`.

"memoized selector" makes sense. However, since we have "input selectors" and if we have an "input" it makes sense that we would have an "output" as well. So I still think "output selector" makes the most sense.

Reply:
Also we have 2 types that I would like to remove entirely if that's okay with you. `ParametricSelector` and `OutputParametricSelector` since `Selector` and `OutputSelector` support additional parameters.

Reply:
Yeah, go ahead and remove those. Breaking change, we'll have to document that, but might as well.

Per the "input/output" thing: I think I disagree because the "input selectors" are "selectors" (args in, derived value out even if it's usually just an extraction), and "inputs to `createSelector`".  The "result function" doesn't get any of the original args, so even though I've _referred_ to it as an "output selector" in the past, that's really not a good usage and I shouldn't have called it that.

The actual generated selector isn't just about "output", though.  It has all three pieces: the input arguments, the memoization, and calculating the final result.  It's not _just_ about "output".  The overall signature is `(...args) => derivedValue`, so it's a standard selector overall, it just happens to be a "memoized" selector that was "generated" by `createSelector` rather than hand-written.  So, I do think that "memoized" or "generated" are the best terms here.

Reply:
Yeah the "result function" is not an "output selector". I agree. That was one of the things that confused me when I first started working on the codebase. the "result function" was sometimes referred to as "the combiner" (which I think in terms of semantics makes the most sense). But it was occasionally referred to as the "output selector", but the genereated selector was also sometimes referred to as the "output selector". In the JSDocs I always called the final generated selector "a memoized output selector". 

BTW I did a quick search in github on `ParametricSelector` and `OutputParametricSelector` seems like anybody was hardly using them so I think we'll be okay removing them, I will go ahead and document their removal.

I still think in terms of simplicity "output selector" makes sense. The "memoized" part is implied and kind of a given. Input -> result function -> output
 sums up the flow of code as well. It's self explanatory in that Reselect goes through the "inputs" then "resultFunc" and gives an "output". It's linear and easier to digest no?